src/DynamicRupture/FrictionLaws/GpuImpl/LinearSlipWeakening.h

```diff
@@ -183,55 +172,167 @@ class LinearSlipWeakeningBase : public GpuFrictionSolver<LinearSlipWeakeningBase
-};
+    auto layerSize{this->currLayerSize};
+    auto deltaT{this->deltaT[timeIndex]};
+    auto* initialStressInFaultCS{this->initialStressInFaultCS};
+    auto* slipRateMagnitude{this->slipRateMagnitude};
+    auto* cohesion{this->cohesion};
+    auto* mu{this->mu};
 
-class LinearSlipWeakeningLawForcedRuptureTime : public LinearSlipWeakeningLaw {
-  public:
-  LinearSlipWeakeningLawForcedRuptureTime(dr::DRParameters& drParameters)
-      : LinearSlipWeakeningLaw(drParameters){};
+    #pragma omp target teams loop              \
+    is_device_ptr(faultStressesPtr,            \
+                  strengthBuffer,              \
+                  initialStressInFaultCS,      \
+                  slipRateMagnitude,           \
+                  cohesion,                    \
+                  mu)                          \
+    device(deviceId) nowait
+    for (unsigned ltsFace = 0; ltsFace < layerSize; ++ltsFace) {
+      auto& faultStresses = faultStressesPtr[ltsFace];
+      auto& strength = strengthBuffer[ltsFace];
 
-  void preHook(real (*stateVariableBuffer)[misc::numPaddedPoints]);
-  void copySpecificLtsDataTreeToLocal(seissol::initializers::Layer& layerData,
-                                      seissol::initializers::DynamicRupture* dynRup,
-                                      real fullUpdateTime);
+      #pragma omp loop bind(parallel)
+      for (unsigned pointIndex = 0; pointIndex < misc::numPaddedPoints; pointIndex++) {
+        //-------------------------------------
+        // calculate Fault Strength
+        // fault strength (Uphoff eq 2.44) with addition cohesion term
+        real totalNormalStress = initialStressInFaultCS[ltsFace][pointIndex][0] +
+                                 faultStresses.normalStress[timeIndex][pointIndex];
+        strength[pointIndex] =
+            -cohesion[ltsFace][pointIndex] -
+            mu[ltsFace][pointIndex] * std::min(totalNormalStress, static_cast<real>(0.0));
+
+        SpecializationT::strengthHook(strength[pointIndex],
+                                      slipRateMagnitude[ltsFace][pointIndex],
+                                      totalNormalStress,
+                                      mu[ltsFace][pointIndex],
+                                      deltaT,
+                                      ltsFace,
+                                      pointIndex);
+      }
+    }
+  }
+
+  void calcStateVariableHook(real (*stateVariableBuffer)[misc::numPaddedPoints],
+                             unsigned int timeIndex) {
+
+    auto layerSize{this->currLayerSize};
+    auto* accumulatedSlipMagnitude{this->accumulatedSlipMagnitude};
+    auto* slipRateMagnitude{this->slipRateMagnitude};
+    auto* forcedRuptureTime{this->forcedRuptureTime};
+    auto* dC{this->dC};
+    auto* resample{this->resampleMatrix};
+    auto deltaT{this->deltaT[timeIndex]};
+    real tn = this->mFullUpdateTime + deltaT;
```

Comment:
would it make sense to try to converge both implementations and add mFullUpdateTime to this in the CPU implementation?

Reply:
I don't think that it is good time to make a completely portable CPU/GPU solutions. It is going to be very messy. Moreover, even now, we need different OpenMP GPU clauses for different compilers (OpenMP 5.0 is not stable across various compiler collections).

The GPU implementation also requires some extra reserved pieces of memory which is redundant for the CPU implementation. All in all, it may look the same by the parallelism is different

Reply:
I meant trying to avoid unnecessarily diverging implementations (to facilitate maintainability), but I get your point.

Reply:
I understood you. I know that it would be much better to get a single code base. Otherwise, one need to always keep in my to fix a bug in two places.

I am going to add `tpv5` and `tpv16` to a GPU CI pipeline. I guess it is the best that I can do for now. Maybe CPU/GPU performance portability in a single code base is going to be a dedicated research topic